src/unchecked-uninit.md

```diff
@@ -8,25 +8,76 @@ Unfortunately this is pretty rigid, especially if you need to initialize your
 array in a more incremental or dynamic way.
 
 Unsafe Rust gives us a powerful tool to handle this problem:
-[`mem::uninitialized`][uninitialized]. This function pretends to return a value
-when really it does nothing at all. Using it, we can convince Rust that we have
-initialized a variable, allowing us to do trickier things with conditional and
-incremental initialization.
-
-Unfortunately, this opens us up to all kinds of problems. Assignment has a
-different meaning to Rust based on whether it believes that a variable is
-initialized or not. If it's believed uninitialized, then Rust will semantically
-just memcopy the bits over the uninitialized ones, and do nothing else. However
-if Rust believes a value to be initialized, it will try to `Drop` the old value!
-Since we've tricked Rust into believing that the value is initialized, we can no
-longer safely use normal assignment.
-
-This is also a problem if you're working with a raw system allocator, which
-returns a pointer to uninitialized memory.
-
-To handle this, we must use the [`ptr`] module. In particular, it provides
-three functions that allow us to assign bytes to a location in memory without
-dropping the old value: [`write`], [`copy`], and [`copy_nonoverlapping`].
+[`MaybeUninit`]. This type can be used to handle memory that has not been fully
+initialized yet.
+
+With `MaybeUninit`, we can initialize an array element-for-element as follows:
+
+```rust
+use std::mem::{self, MaybeUninit};
+
+// Size of the array is hard-coded but easy to change. This means we can't
+// use [a, b, c] syntax to initialize the array, though!
+const SIZE: usize = 10;
+
+let x = {
+    // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
+    // safe because the type we are claiming to have initialized here is a
+    // bunch of `MaybeUninit`s, which do not require initialization.
+    let mut x: [MaybeUninit<Box<u32>>; SIZE] = unsafe {
+        MaybeUninit::uninit().assume_init()
+    };
+
+    // Dropping a `MaybeUninit` does nothing. Thus using raw pointer
+    // assignment instead of `ptr::write` does not cause the old
+    // uninitialized value to be dropped.
+    // Exception safety is not a concern because Box can't panic
+    for i in 0..SIZE {
+        x[i] = MaybeUninit::new(Box::new(i as u32));
+    }
+
+    // Everything is initialized. Transmute the array to the
+    // initialized type.
+    unsafe { mem::transmute::<_, [Box<u32>; SIZE]>(x) }
```

Comment:
So, what do you propose I change here? Just add more comments saying `DANGER WARNING`?

Reply:
Yeah, but we should also have at least one or two paragraphs dedicated to like "hey this is a thing" because it's the most common thing I see people asking about, hoping they can just do it.

Also if your type isn't repr(C), adding MaybeUninit technically can lead to layout randomization for no reason other than the compiler being capricious.

Reply:
`MaybeUninit` is `repr(transparent)` -- it is not entirely clear what that means for unions, but transmute is definitely allowed.

> Yeah, but we should also have at least one or two paragraphs dedicated to like "hey this is a thing" because it's the most common thing I see people asking about, hoping they can just do it.

I don't follow. What is a thing?

Reply:
Pushed something. Let me know what you think.